src/textual/widgets/text_input.py

```diff
@@ -222,11 +253,22 @@ def on_key(self, event: events.Key) -> None:
             # window/range along to the right.
             if cursor_index == end - 1:
                 self.visible_range = start + 1, end + 1
+            self._update_suggestion(event)
 
         # We need to clamp the visible range to ensure we don't use negative indexing
         start, end = self.visible_range
         self.visible_range = (max(0, start), end)
 
+    def _update_suggestion(self, event: events.Key) -> None:
+        """Run the autocompleter function, updating the suggestion if necessary"""
+        if self.autocompleter is not None:
+            event.prevent_default()
+            super().on_key(event)
+            if self.value:
+                self._suggestion = self.autocompleter(self.value)
```

Comment:
I guess we could have something like this here, so we can handle both sync and async autocompleters? :-)
```py
if self.value:
    if isawaitable(self.autocompleter):
        self._suggestion = await self.autocompleter(self.value)
    else:       
        self._suggestion = self.autocompleter(self.value)
``` 

...which implies that `_update_suggestion` must become async, and so must `on_key` as it seems to be the function that calls it? 
But if I'm not wrong the Textual events management already handles async event managers, using the same `if isawaitable(...)` trick, so it shouldn't be an issue? :slightly_smiling_face: